Exposed Control::focus_mode to the scene editor

leezh opened the pull request:
Allows fine-tuning of tab focusing without using scripts.

Exposed `get_focus_mode()` to the script
Added `focus_mode` to the property panels for line-edit and sliders
Added `enabled_focus_mode` to the property panels for buttons

enabled_focus_mode is used when button is enabled/disabled

leezh committed:
GUI Focus mode improvements

Exposed `get_focus_mode()` to the script
Added `focus_mode` to the property panels for line-edit and sliders
Added `enabled_focus_mode` to the property panels for buttons

enabled_focus_mode is used when button is enabled/disabled

leezh commented:
Actually, hold on one sec. I realised that if you replace a non-interactive control with something like a button, this property gets carried over.

leezh commented:
Perhaps I should only expose it on controls that actually use it?

akien-mga commented:
Any update @leezh? :)

neikeq commented:
@leezh I don't understand your last comment. What's wrong with the property getting carried? AFAIK, all Control derived classes can change its focus mode.

leezh commented:
Basically, when you create a control, like a frame, the initial property is FOCUS_NONE. Later on, if you use the replace node tool to change it to say, a button, that property is still going to be FOCUS_NONE. It's not too bad if you remember to change it, but there's no way a beginer would've noticed it. In fact, it's annoying that you *have* to change it every time you replace a node. And that's my issue.

I've been trying to get it to be like one of those custom theme properties (where you can check and uncheck whether the property is actually saved) but I can't get it to work right.

neikeq commented:
IMO, it's fine that way.

akien-mga commented:
Resetting properties to their default value when changing the node type is a more global issue IMO, it also happens for signals for example. As long as it's not a regression, it should not be blocking for this PR.

akien-mga commented:
The last commit seems not to build:
```
scene/gui/base_button.cpp:380:29: error: variable or field 'set_enabled_focus_mode' declared void
scene/gui/base_button.cpp:380:29: error: 'FocusMode' was not declared in this scope
```
(see travis logs)

akien-mga commented:
The PR needs rebasing, and still fails to build.

leezh commented:
Ah, sorry. Was busy with my thesis until last week. Haven't had a time to sit down and fix this. It was a `git pull --rebase` that went horribly wrong.

leezh commented:
Fixed the problems I made in the last commit and squashed them all together. So what's changed now is that the property is exposed only to a few control types that makes sense.

Also, because of how the BaseButton can switch back and forth when enabling/disabling, I found it far more useful to have an enabled_focus_mode property instead.
